modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -785,27 +793,39 @@ public void onError(Throwable e) {
     }
 
     /**
-     * Saves states of the Distribution Zone Manager to vault atomically in one batch.
+     * Update topology augmentation map with newly added and removed nodes.
+     *
+     * @param addedNodes Nodes that was added to a topology and should be added to zones data nodes.
+     * @param removedNodes Nodes that was removed from a topology and should be removed from zones data nodes.
+     * @param revision Revision of the event that triggered this method.
+     * @param zoneId Zone's id.
+     */
+    private void updateTopologyAugmentationMap(Set<Node> addedNodes, Set<Node> removedNodes, long revision, int zoneId) {
+        if (!addedNodes.isEmpty()) {
+            zonesState.get(zoneId).nodesToAddToDataNodes(addedNodes, revision);
+        }
+
+        if (!removedNodes.isEmpty()) {
+            zonesState.get(zoneId).nodesToRemoveFromDataNodes(removedNodes, revision);
+        }
+    }
+
+    /**
+     * Saves states of the Distribution Zone Manager to Meta Storage atomically in one batch.
      * After restart it could be used to restore these fields.
      *
-     * @param newLogicalTopology Logical topology.
-     * @param revision Revision of the event, which triggers update.
      * @param zoneIds Set of zone id's, whose states will be staved in the Vault
      */
-    private void saveStatesToVault(Set<NodeWithAttributes> newLogicalTopology, long revision, Set<Integer> zoneIds) {
-        Map<ByteArray, byte[]> batch = IgniteUtils.newHashMap(3 + zoneIds.size());
+    private CompletableFuture<Void> saveStatesToMetastorage(Set<Integer> zoneIds) {
```

Comment:
Add returned value in javadoc with description when this future will be completed.